crypto/bio/bio_lib.c

```diff
@@ -848,35 +857,44 @@ int BIO_connect_retry(BIO *bio, int timeout)
         return 0;
     }
 
+    if (polling_milliseconds < 0)
+        polling_milliseconds = 100;
     if (!blocking)
         BIO_set_nbio(bio, 1);
 
- retry: /* it does not help here to set SSL_MODE_AUTO_RETRY */
-    rv = BIO_do_handshake(bio); /* This indirectly calls ERR_clear_error(); */
+ retry: /* it does not always help to set SSL_MODE_AUTO_RETRY */
+    rv = BIO_do_connect(bio); /* This may indirectly call ERR_clear_error(); */
 
     if (rv <= 0) {
-        if (get_last_sys_error() == ETIMEDOUT) {
-            /*
-             * if blocking, despite blocking BIO, BIO_do_handshake() timed out
-             * when non-blocking, BIO_do_handshake() timed out early
-             * with rv == -1 and get_last_sys_error() == 0
-             */
-            ERR_clear_error();
-            (void)BIO_reset(bio);
-            /*
-             * unless using BIO_reset(), blocking next connect() may crash and
-             * non-blocking next BIO_do_handshake() will fail
-             */
-            goto retry;
-        } else if (BIO_should_retry(bio)) {
-            /* will not actually wait if timeout == 0 (i.e., blocking BIO) */
-            rv = bio_wait(bio, max_time, 100 /* milliseconds */);
+        int err = ERR_peek_last_error();
+        int reason = ERR_GET_REASON(err);
+        int do_retry = BIO_should_retry(bio);
+
+        if (ERR_GET_LIB(err) == ERR_LIB_BIO) {
+            if (reason == BIO_R_NBIO_CONNECT_TIMEOUT
+                    || reason == BIO_R_CONNECT_TIMEOUT) {
```

Comment:
With the BIO_R_CONNECT_TIMEOUT this is really an endless loop.
the blocking connect is only tried when timeout == 0
I would also not differentiate between NBIO_CONNECT_TIMEOUT and NBIO_CONNECT_ERROR,
if the server does not listen, you can get a CONNECT_ERROR and you can of course
try that connection again after a short nap.

Reply:
> With the BIO_R_CONNECT_TIMEOUT this is really an endless loop.
the blocking connect is only tried when timeout == 0

Indeed it could result in an endless loop (yet not in a busy loop, due to `BIO_wait`),
but if the user sets `timeout == 0` this is exactly what he wants: keep trying to connect as long as whatever needed (even infinitely) as long as the connection cannot be established.

> I would also not differentiate between NBIO_CONNECT_TIMEOUT and NBIO_CONNECT_ERROR,
if the server does not listen, you can get a CONNECT_ERROR and you can of course
try that connection again after a short nap.

Interesting thought. Yet I would assume that `NBIO_CONNECT_ERROR` is something fatal, while `NBIO_CONNECT_TIMEOUT` according to my experiments is a non-fatal error just indicating that within the strange 30 s internal grace period no connection could be established, so when the user is patient enough it makes sense to retry.

Reply:
It is your function.

But again, I did not assume that timeout=0 means forever, I did just assume,
try the connection, to all addresses that match to the request exactly once.

For the blocking connect, you get only the error code of the last connect
attempt, the first few could have a CONNECT_ERROR and then a few
CONNECT_TIMEOUT and then it is an endless loop.
But if the first connect has a CONNECT_TIMEOUT and then a CONNECT_ERROR,
then you stop the loop.

Reply:
Same in the non-blocking connect.
You get here only when the *last* address that getaddrinfo returned
had a TIMEOUT.  That looks bogus to take that as an indication
when to terminate the loop.

Reply:
You just convinced me that it is more consistent - and also simpler of course - not to distinguish the timeout and error cases. 
I'll give that a try...

Reply:
So I've unified the handling of  CONNECT_TIMEOUT and CONNECT_ERROR,
which is most likely an improvement, at least in terms of consistency.

It turns out that this does not yet cover the system error cases, in particular `EAGAIN` of `getaddrinfo` (Temporary failure in name resolution), so I've added the `ERR_R_SYS_LIB` case to cover all system errors, assuming that it makes sense to retry on any of them.

Experiments after these changes went fine. 
So it looks like we really can treat all non-fatal errors like internal  timeouts: reset the BIO, clear the error queue, and retry.